File: Data Structures & Algorithms/eating-bananas/submission-1.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l = 1
        r = max(piles)

        while l < r:
            mid = (l + r) // 2
            hours = sum((p + mid - 1) // mid for p in piles)

            if hours <= h:
                r = mid
            else:
                l = mid + 1

        return l

File: Data Structures & Algorithms/capacity-to-ship-packages-within-d-days/submission-1.ts
class Solution {
    /**
     * @param {number[]} weights
     * @param {number} days
     * @return {number}
     */
    shipWithinDays(weights: number[], days: number): number {
        let l: number = Math.max(...weights);
        let r: number = weights.reduce((a, b) => a + b, 0);

        while (l < r) {
            let mid = Math.floor((l + r) / 2);
            let dayNeeded: number = 1;
            let curLoad: number = 0;

            for (let w of weights) {
                if (curLoad + w > mid) {
                    dayNeeded++;
                    curLoad = 0;
                }

                curLoad += w;
            }

            if (dayNeeded <= days) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }

        return l;
    }
}
